remoulade/state/backends/postgres.py

```diff
@@ -125,29 +157,59 @@ def get_states(
         end_datetime: Optional[datetime.datetime] = None,
         sort_column: Optional[str] = None,
         sort_direction: Optional[str] = None,
-        get_groups: bool = False,
+        selected_type: Optional[str] = None,
     ):
+        sort_column = sort_column or "enqueued_datetime"
+        sort_direction = sort_direction or "desc"
+
         with self.client.begin() as session:
             query = session.query(StoredState)
-            if selected_actors is not None:
-                query = query.filter(StoredState.actor_name.in_(selected_actors))
-            if selected_statuses is not None:
-                query = query.filter(StoredState.status.in_(selected_statuses))
-            if selected_ids is not None:
-                query = query.filter(StoredState.message_id.in_(selected_ids))
-            if start_datetime is not None:
-                query = query.filter(StoredState.enqueued_datetime >= start_datetime)
-            if end_datetime is not None:
-                query = query.filter(StoredState.enqueued_datetime <= end_datetime)
-            if get_groups:
-                query = query.filter(StoredState.group_id.is_not(None))
-            if sort_column in State._fields and sort_direction is not None:
-                query = query.order_by(text(f"{sort_column} {sort_direction}"))
+            query = filter_query(
+                query, selected_actors, selected_statuses, selected_ids, start_datetime, end_datetime, selected_type
+            )
+            query = query.order_by(text(f"{sort_column} {sort_direction}"))
             if size is not None:
-                query = query.offset(offset).limit(size)
+                query = query.subquery()
+                query_group = session.query(
+                    functions.max(query.c.composition_id).label("composition_id"),
+                    functions.max(query.c.message_id).label("message_id"),
+                ).group_by(coalesce(query.c.composition_id, query.c.message_id))
+                query_group = query_group.offset(offset).limit(size).subquery()
+                query = (
+                    session.query(StoredState)
+                    .select_from(StoredState)
+                    .join(
+                        query_group,
+                        or_(
+                            StoredState.message_id == query_group.c.message_id,
+                            StoredState.composition_id == query_group.c.composition_id,
+                        ),
+                    )
+                )
 
             return [state_model.as_state(self.encoder) for state_model in query]
```

Comment:
```suggestion
            return (state_model.as_state(self.encoder) for state_model in query)
```
to return an iterator, otherwise you are looping twice on it when you do `as_dict` in the api